helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -1036,6 +1037,46 @@ public static void purgeExpiredJobs(String workflow, WorkflowConfig workflowConf
     setNextJobPurgeTime(workflow, currentTime, purgeInterval, rebalanceScheduler, manager);
   }
 
+  /**
+   * The function that loops through the all existed workflow contexts and removes IdealState and
+   * Workflow Context if Workflow Config is missing.
+   * @param dataProvider
+   * @param manager
+   */
+  public static void workflowGarbageCollection(WorkflowControllerDataProvider dataProvider,
```

Comment:
Here we are doing a loop based on workflow contexts. Could there be cases where IdealStates exist but contexts/configs do not?

Reply:
Since controller attempts to delete the workflow context after the successful deletion of the workflow config and IdealState, there is no need to consider cases where IS existed without contexts/configs.

Reply:
Let's consider two scenarios: 

- Case 1 where workflow config has been deleted but workflow context and IS have not been removed. In this case this PR addresses this scenario and deletes both of them. Note that in this PR context won't be removed unless IS is removed. If IS has not been successfully removed, we will try to delete it again in next pipeline.

- Case 2 where workflow config and context have been removed and IS has not been removed. I am arguing that this is not possible since context is always the last one that controller removes. 

PS: I don't know any way that controller can find all of the IdealStates that are associated to workflows and not the regular resources (we can do that for context based on the Znode ID). If we have such method, we can also loop through the IdealStates in addition to contexts.